### PR TITLE
server : Smart selection of available slot using Longest Common Prefix

### PR DESCRIPTION
**Pull request by sasha0552**
In the current implementation, an available slot is selected using LRU (Least Recently Used). This PR adds slot selection by ~~[LCS](https://en.wikipedia.org/wiki/Longest_common_substring) (Longest Common Substring)~~ LCP (Longest Common Prefix) algorithm to select a slot with a prompt that has at least `n%` similarity to the requested prompt. This reduces prompt processing in multi-user scenarios.

Additionally, this PR:
1. Defers processing if a slot is unavailable and the user has explicitly requested a specific slot.
2. Fixes erase/save/load slot endpoints. In the current implementation, these endpoints use a slot by LRU if the requested slot is not available, which is not correct behavior.

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]
Need to rebase to latest `master` and we can merge

File: common/common.cpp [hunk 1]
```diff
@@ -1460,6 +1460,14 @@ bool gpt_params_find_arg(int argc, char ** argv, const std::string & arg, gpt_pa
...
         return true;
     }
+    if (arg == "--lcs-similarity") {
+        if (++i >= argc) {
+            invalid_param = true;
+            return true;
+        }
+        params.lcs_similarity = std::stof(argv[i]);
+        return true;
+    }
```
Comment (ggerganov) [inline common/common.cpp:1470]: Add usage entry below in the "server" section

**Comment by sasha0552** [comment]
I'll test again and then mark the PR as ready for review.

**Comment by sasha0552** [comment]
By the way, should this be on by default? Or is it better to leave it off as it is now?

**Comment by ggerganov** [comment]
> This PR adds slot selection by [LCS](https://en.wikipedia.org/wiki/Longest_common_substring) (Longest Common Substring) algorithm to select a slot with a prompt that has at least n% similarity to the requested prompt. This reduces prompt processing in multi-user scenarios.

The LCS algorithm is an overkill for this purpose. All you need to look for is the longest common prefix, which is much simpler to compute

**Comment by sasha0552** [comment]
As far as I know, the server can reuse not only the prompt prefix, but also the suffix (`llama_kv_cache_seq_rm` & `llama_kv_cache_seq_add sequence`).

https://github.com/ggerganov/llama.cpp/blob/2b3389677a833cee0880226533a1768b1a9508d2/examples/server/server.cpp#L1819-L1857

**Comment by ggerganov** [comment]
Although the `llama` interface allows various operations with the KV cache, such as shifting the tokens' position and removing tokens, the `server` only looks for a common prefix when the prompt cache is enabled:

https://github.com/ggerganov/llama.cpp/blob/d5c938cd7716b9a2ace49a43a469dfbffcff4d28/examples/server/server.cpp#L2022-L2025

So for now the slot selection logic is better to follow the prompt caching logic and look just at the prefix

**Comment by samikrc** [comment]
Would this be adding new parameters to the server command line in order to enable this feature?

**Pull request merged**